mimic/test/behavior_tests.py

```diff
@@ -0,0 +1,242 @@
+"""
+Automatically generate tests for behavior registration/deletion APIs.
+Also contains helper functions for specific behavior testing.
+"""
+import json
+
+from uuid import UUID, uuid4
+
+from six import text_type, string_types
+
+from twisted.trial.unittest import SynchronousTestCase
+
+from zope.interface import Attribute, Interface
+
+from mimic.test.helpers import json_request, request_with_content
+
+
+class IBehaviorAPITestHelperFactory(Interface):
+    """
+    A class with a `classmethod` that accepts a test case and returns a
+    helper.  This will be used to generate a behavior CRUD test suite.
+    """
+    __name__ = Attribute(
+        "A name which will be used to generate the test suite name.")
+    __module__ = Attribute(
+        "The module in which the test suite should go.")
+
+    def from_test_case(test_case):
+        """
+        A constructor that generates a provider of
+        :class:`IBehaviorAPITestHelper`.
+
+        :param test_case: An instance of a
+            :class:`twisted.trial.unittest.SynchronousTestCase`
+        :return: a :class:`IBehaviorAPITestHelper` provider
+        """
+
+
+class IBehaviorAPITestHelper(Interface):
+    """
+    Set-up and assertion methods required for testing behavior CRUD.
+    :class:`IBehaviorAPITestHelperFactory.from_test_case` will be called at
+    the beginning of every CRUD method, the intent of which is to generate
+    a new instance should this be a class.  If the provider is not a class,
+    then please keep in mind the intended usage.
+
+    The reason there are two interfaces, both clearly intended to be
```

Comment:
The reason that there are two interfaces is that there are two separate responsibilities.  One thing creates a `Helper` thing, the other thing provides the behavior of that `Helper` thing.  Diamond inheritance, while always a nasty edge-case, is not the only reason to avoid inheritance.  For example, if Trial itself had a clean implementer/provider distinction around how test cases are discovered and instantiated, rather than requiring discoverable test suites to be literal class objects inheriting from an `ITestCase` implementor, this code would be a lot simpler to implement and would be using test-suite APIs rather than special attributes like `__name__` and `__module__`.

I gather that you added this comment in response to the fact that factory-pattern interfaces are a bit more verbose to document with zope.interface than other kinds of interface, because you need a class/class docstring/method/method docstring as opposed to just class/class docstring as with a class.  However, documenting creation patterns with a clear distinction between what is expected to be provided by the creator and what is expected to be provided by the createe is a worthwhile habit to get into.

So rather than saying that they are "clearly intended to be implemented by a single class", since there's nothing special about this interface requiring it to be one class any more than any other interface-that-returns-something, would be to have a single high-level API for the common case of using a single class, as I recommended in my previous comment. It's a little more verbosity in the implementation, but in all of the API clients, it turns multiple lines of boilerplate into a one-liner which can be clearly documented.